### ISSUE TITLE
Why TensorRT use Convolution instead MatMul in explicit quantized model

### ISSUE THREAD

**Issue by WoodieDudy** [NVIDIA/TensorRT#3266]
I have model
```python
class MLP(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        d_model, d_ff = 512, 2048
        self.lin1 = nn.Linear(d_model, d_ff)
        self.activation = nn.ReLU()
        self.lin2 = nn.Linear(d_ff, d_model)

    def forward(self, x):
        x = self.lin1(x)
        x = self.activation(x)
        x = self.lin2(x)
        return x
```
I exported it to ONNX using explicit quantization with `pytorch_quantization`. 

```python
import torch
from pytorch_quantization import nn as quant_nn
from pytorch_quantization import quant_modules

quant_nn.TensorQuantizer.use_fb_fake_quant = True
quant_modules.initialize()
model = MLP().eval()
    
torch.onnx.export(
    model.cuda(),
    torch.rand(10240, 512).cuda(),
    "MLP_explicit_quant_fp32.onnx",
    verbose=False,
    input_names=["x"],
    opset_version=17
)
```
[MLP_explicit_quant_fp32.onnx.zip](https://github.com/NVIDIA/TensorRT/files/12431552/MLP_explicit_quant_fp32.onnx.zip)

To build and visualize model
```
python TensorRT/tools/experimental/trt-engine-explorer/utils/process_engine.py MLP_explicit_quant_fp32.onnx temp int8 fp16
```
Output with trtexec commands to build same engine:
```
Building the engine:
trtexec --verbose --nvtxMode=verbose --buildOnly --workspace=8192 --onnx=onnx/MLP_explicit_quant_fp32.onnx --saveEngine=temp/MLP_explicit_quant_fp32.onnx.engine --timingCacheFile=./timing.cache --int8 --fp16

Successfully built the engine.

Engine building metadata: generated output file temp/MLP_explicit_quant_fp32.onnx.engine.build.metadata.json
Profiling the engine:
trtexec --verbose --noDataTransfers --useCudaGraph --separateProfileRun --useSpinWait --nvtxMode=verbose --loadEngine=temp/MLP_explicit_quant_fp32.onnx.engine --exportTimes=temp/MLP_explicit_quant_fp32.onnx.engine.timing.json --exportProfile=temp/MLP_explicit_quant_fp32.onnx.engine.profile.json --exportLayerInfo=temp/MLP_explicit_quant_fp32.onnx.engine.graph.json --timingCacheFile=./timing.cache --int8 --fp16
WARNING:root:Could not lock clocks (Insufficient Permissions).
	Try running as root or locking the clocks from the commandline:
		sudo nvidia-smi --lock-gpu-clocks=1410,1410
		sudo nvidia-smi --applications-clocks=1215,1410
WARNING:root:Could not unlock clocks (Insufficient Permissions).
	Try running as root or unlocking the clocks from the commandline:
		sudo nvidia-smi --reset-gpu-clocks
		sudo nvidia-smi --reset-applications-clocks

Successfully profiled the engine.

Profiling metadata: generated output file temp/MLP_explicit_quant_fp32.onnx.engine.profile.metadata.json
Generating graph diagram: temp/MLP_explicit_quant_fp32.onnx.engine.graph.json
/root/projects/TensorRT/tools/experimental/trt-engine-explorer/trex/engine_plan.py:90: UserWarning:

Profiling data was not provided.

Created file:///root/projects/TensorRT/tools/experimental/trt-engine-explorer/temp/MLP_explicit_quant_fp32.onnx.engine.graph.json.svg
Artifcats directory: temp
```
Build logs:
[MLP_explicit_quant_fp32.onnx.engine.build.log](https://github.com/NVIDIA/TensorRT/files/12431339/MLP_explicit_quant_fp32_whyfp32.onnx.engine.build.log)

### I noticed that TensorRT use Convolution instead MatMul despite the fact that @nvpohanh said [here](https://github.com/NVIDIA/TensorRT/issues/3233?notification_referrer_id=NT_kwDOAA_lWLI3NDE1ODg5Mjg5OjEwNDE3NTI#issuecomment-1689532096) that with explicit quantization Convilution should be replaced by MatMul

![MLP_explicit_quant_fp32 onnx engine graph json](https://github.com/NVIDIA/TensorRT/assets/55787608/6bc1671f-82ab-4883-a8ed-cbb31e768602)

**Comment by nvpohanh** [comment]
By default, TRT uses INT8 Convs for all the INT8 Gemms. In theory, they should have similar performance.

If you really want to use INT8 Gemm kernels, please add a LayerNormalization or an MHA pattern to trigger the Transformer optimization in TRT.

**Comment by nvpohanh** [comment]
Yes, that is expected. So now if you enable explicit quantization (by adding Q/DQ ops before the gemm), you should see INT8 gemm kernels being used.

**Comment by WoodieDudy** [comment]
I already have Q/DQ added as you can see in the first svg picture. Does this mean that the calculations inside myelin are done in int8?
Original onnx looks like this
<img width="322" alt="image" src="https://github.com/NVIDIA/TensorRT/assets/55787608/3c4150c8-fd45-4357-b84e-2c89782e4884">

**Comment by nvpohanh** [comment]
I think we will need to check the Nsys profile to know exactly which kernels are used.

**Comment by WoodieDudy** [comment]
I profiled it. So it uses int8 inside👍
![SCR-20230904-oozi](https://github.com/NVIDIA/TensorRT/assets/55787608/634d65f9-b267-4b21-bf03-732c6694c4e7)

**Comment by nvpohanh** [comment]
That kernel (`igemm_int8`) is INT8-input FP32-output gemm kernel and it does not use TensorCores. Could you add additional Q/DQ after the gemm so that it uses INT8-output to achieve potentially better perf? Thanks

**Comment by nvpohanh** [comment]
I don't think IO formats matters, though. You just need to add Q/DQ after the last gemm

**Comment by vadimkantorov** [comment]
well, the output of such MLP transformer block might not be actually quantized, rather a simple fp16 output is needed...

**Comment by phantaurus** [comment]
> By default, TRT uses INT8 Convs for all the INT8 Gemms. In theory, they should have similar performance.

Hello! @nvpohanh 

I think Conv kernels require activation input to have NC/xHWx layout, which could lead to the insertion of extra transpose/reformat operations.

Specifically, if I write a custom NHWC kernel/plugin (e.g., FlashAttn) before the Conv kernel, TensorRT will first convert NHWC to NCHW, then convert NCHW to NC/xHWx, before actually calling the Conv kernel. If I have another NHWC kernel following the Conv kernel, TensorRT will have to insert two more transposes after the Conv.

I understand that if I have a chain of Conv layers, then the layout will be NC/xHWx throughout. The transpose only happens at the beginning and the end of the chain. However, if I want to mix up the usage of Conv and Matmul, there will constantly be format adjustments. This will introduce significant slowdown if my entire model is float or fp16.

However, I do notice that this transpose operation is less of an issue when, say, I naturally need to cast float values to int8 before doing int8 CastConvolution. It seems that the cast operation and the reformat operation are fused in some sense, and the overhead is not too different from doing cast alone. Is this the reason why TensorRT prefers CastConvolution implementation for all int8 GEMMs?

=========

That being said, I'm guessing that if I have a full int8 model, in which all layers have int8 inputs and int8 outputs, then the transpose/reformat cost might again become non-negligible?

But I’m actually having issues creating such a full int8 model, which leads to my second question regarding the fusion of int8 CastConvolution and Relu. In most cases, TensorRT (8.5) breaks a fake-quantized torch.nn.Linear layer into two layers: one does int8 CastConvolution, the other is the fusion of BiasAdd and Relu.

The CastConvolution layer has int8 input and fp16 output, while the fused element-wise layer has fp16 inputs and outputs. I'm wondering if BiasAdd and Relu can also be performed in fixed-point? Theoretically, it seems possible. If that is possible, a chain of torch.nn.Linear could be full int8. However, right now it seems that TensorRT 8.5 is giving int8 CastConvolution with fp16 output, then fp16 element-wise, then casting to int8, then int8 CastConvolution with fp16 output again.

Thanks!

**Comment by vadimkantorov** [comment]
@nvpohanh It would be extremely helpful to have complete end-to-end examples with PyTorch code snippets and `trtexec` command examples, along with trex visualizations, profiler outputs, perf improvements, measurements

At least for several typical models, including MLPs, BERTs, ConvNets etc
